lib/rules/no-misleading-character-class.js

```diff
@@ -371,12 +381,31 @@ module.exports = {
                 for (const { node: refNode } of tracker.iterateGlobalReferences({
                     RegExp: { [CALL]: true, [CONSTRUCT]: true }
                 })) {
+                    let pattern, flags;
                     const [patternNode, flagsNode] = refNode.arguments;
-                    const pattern = getStringIfConstant(patternNode, scope);
-                    const flags = getStringIfConstant(flagsNode, scope);
+                    const evaluatedPattern = getStaticValue(patternNode, scope);
+
+                    if (!evaluatedPattern) {
+                        continue;
+                    }
+                    if (flagsNode) {
+                        if (isRegExp(evaluatedPattern.value)) {
+                            pattern = evaluatedPattern.value.source;
+                            checkedPatternNodes.add(patternNode);
+                        } else {
+                            pattern = String(evaluatedPattern.value);
+                        }
+                        flags = getStringIfConstant(flagsNode, scope);
```

Comment:
Can you add this test:

```js
{
    code: "new RegExp(/^[👍]$/v, '')",
    languageOptions: {
        ecmaVersion: 2024
    },
    errors: [{
        column: 15,
        endColumn: 17,
        messageId: "surrogatePairWithoutUFlag",
        suggestions: [{ messageId: "suggestUnicodeFlag", output: "new RegExp(/^[👍]$/v, 'u')" }]
    }]
}
```

It's interesting to see what will happen on Node 18 because it doesn't support the `v` flag (per [node.green](https://node.green/#ES2024-features-RegExp--v--flag))

Reply:
Good catch! This will fail in Node.js 18 because the parser silently sets the node value to `null` if it cannot evaluate it. It's also a known limitation of eslint-utils ([code](https://github.com/eslint-community/eslint-utils/blob/0540eb3435a80cce4a02c663833d800812dcff51/src/get-static-value.mjs#L442-L445)), but I don't think we want to have a different behavior depending on the version of Node.js.

The only way I found to fix this in ESLint directly is by patching the AST with the missing `value`s (see commit ed8d1cd606abc86644a2f6948af23439bd092c82), but this could break third-party rules that don't expect this change, so not sure. Maybe a safer approach would be extending `getStaticValue()` in eslint-utils under an option to keep the current API unchanged.

Reply:
> The only way I found to fix this in ESLint directly is by patching the AST with the missing `value`s (see commit [ed8d1cd](https://github.com/eslint/eslint/commit/ed8d1cd606abc86644a2f6948af23439bd092c82)), but this could break third-party rules that don't expect this change, so not sure. Maybe a safer approach would be extending `getStaticValue()` in eslint-utils under an option to keep the current API unchanged.

I think that's out of scope for this change, but we could evaluate it separately.

As for this particular case, maybe we could additionally check if the first argument is a regex literal? That would retain the current functionality where an error is reported regardless of the Node.js version.

Reply:
> As for this particular case, maybe we could additionally check if the first argument is a regex literal? That would retain the current functionality where an error is reported regardless of the Node.js version.

That would not work as expected when the first argument is a variable, e.g.:

```js
var r = /[👶🏻]/v;  // error
RegExp(r, 'v');   // error
```

In Node.js 18, the second line would not be flagged because `getStaticValue()` reports a value of `null` for the first argument.

Maybe we should just check if the node is a regex literal without using `getStaticValue()` when there is a second parameter?

Reply:
> Maybe we should just check if the node is a regex literal without using `getStaticValue()` when there is a second parameter?

What would the rule check for the RegExp constructors node with two arguments n case the first argument isn't a regex literal?

Reply:
> That would not work as expected when the first argument is a variable, e.g.:
> 
> ```js
> var r = /[👶🏻]/v;  // error
> RegExp(r, 'v');   // error
> ```

In this case, the regex literal would be reported in all Node.js versions, but the RegExp constructor would only in Node.js > 18? That's definitely something we should aim to avoid, but I think this was the case before this change as well so we're at least not introducing new cases where the behavior depends on the runtime.

Reply:
> What would the rule check for the RegExp constructors node with two arguments in case the first argument isn't a regex literal?

If the first argument is not a regex literal we could call `getStaticValue()`, but if that returns a regular expression object then we should ignore it to avoid inconsistencies across different versions of Node.js.

> > That would not work as expected when the first argument is a variable, e.g.:
> > ```js
> > var r = /[👶🏻]/v;  // error
> > RegExp(r, 'v');   // error
> > ```
> 
> In this case, the regex literal would be reported in all Node.js versions, but the RegExp constructor would only in Node.js > 18? That's definitely something we should aim to avoid, but I think this was the case before this change as well so we're at least not introducing new cases where the behavior depends on the runtime.

Yes, if it's okay to keep the existing bug that's also an option.

Reply:
> If the first argument is not a regex literal we could call `getStaticValue()`, but if that returns a regular expression object then we should ignore it to avoid inconsistencies across different versions of Node.js.

Good idea! Let's do that to fix inconsistencies.

At some future point, we could see if `getStaticValue` could be improved to better handle regexes that can't be instantiated, and then update this code.

Reply:
Okay, I've done the changes in 7eaf8b45e4dfe9c0e80d8e3309ad43ba1a57cac0. I had to comment out some of the unit tests that were no longer passing. Maybe we could also update the tests to make them pass and add a note that the expected results are not really what was intended because of the limitations with Node.js 18.

When I have some time I would like to look at how `getStaticValue` is used throughout ESLint and come up to eslint-community with a proposal to extend that API, so we can remove those limitations.